core/metacling/src/TClingCallFunc.cxx

```diff
@@ -1389,384 +1391,85 @@ TClingCallFunc::make_dtor_wrapper(const TClingClassInfo *info)
    return (tcling_callfunc_dtor_Wrapper_t)F;
 }
 
-class ValHolder {
-public:
-   union {
-      long double ldbl;
-      double dbl;
-      float flt;
-      //__uint128_t ui128;
-      //__int128_t i128;
-      unsigned long long ull;
-      long long ll;
-      unsigned long ul;
-      long l;
-      unsigned int ui;
-      int i;
-      unsigned short us;
-      short s;
-      //char32_t c32;
-      //char16_t c16;
-      //unsigned wchar_t uwc; - non-standard
-      wchar_t wc;
-      unsigned char uc;
-      signed char sc;
-      char c;
-      bool b;
-      void *vp;
-   } u;
-};
-
 void TClingCallFunc::exec(void *address, void *ret)
 {
-   SmallVector<ValHolder, 8> vh_ary;
-   SmallVector<void *, 8> vp_ary;
+   const FunctionDecl *FD = GetDecl();
+   const unsigned num_args = fArgVals.size();
+
+   // FIXME: Consider the implicit this which is sometimes not passed.
+   if (num_args < GetMinRequiredArguments()) {
+      ::Error("TClingCallFunc::exec",
+              "Not enough arguments provided for %s (%d instead of the minimum %d)",
+              fMethod->Name(),
+              num_args, (int)GetMinRequiredArguments());
+      return;
+   } else if (!isa<CXXMethodDecl>(FD) && num_args > FD->getNumParams()) {
+      ::Error("TClingCallFunc::exec",
+              "Too many arguments provided for %s (%d instead of the minimum %d)",
+              fMethod->Name(),
+              num_args, (int)GetMinRequiredArguments());
+      return;
+   }
+   if (auto CXXMD = dyn_cast<CXXMethodDecl>(FD))
+     if (!address && CXXMD && !CXXMD->isStatic() && !isa<CXXConstructorDecl>(FD)) {
+       ::Error("TClingCallFunc::exec",
+               "The method %s is called without an object.",
+               fMethod->Name());
+       return;
+     }
 
-   unsigned num_args = fArgVals.size();
+   SmallVector<void *, 8> vp_ary;
    {
       R__LOCKGUARD_CLING(gInterpreterMutex);
```

Comment:
@pcanal, now this loop accesses only properties of the AST which are readonly and pushes things to `vp_ary`. Do you think we still need this one?

Reply:
Are we guaranteed that the AST operations trigger no deserializations? Any loop over decl context elements or overloads might, iirc.

Reply:
We are not looping over AST elements. We query for some already constructed properties of the WualType. Can you be more specific which is the part of the current loop that worries you?

Reply:
const operation here are not good enough to know whether the lock is necessary or not because the 'clang database' is not fine grained locked (at least not in our builds) and thus another thread can modify the database (unload decl or add template instantiation decls for example) while this thread is doing const operations.

Reply:
Ok, then we can keep it, although I do not see any of these changing...

Reply:
possible change: unloading, addition of template instances, etc.